lgc/test/InOutPackingNonZeroBase.lgc

```diff
@@ -0,0 +1,130 @@
+; RUN: lgc -mcpu=gfx900 --stop-after=lgc-patch-resource-collect %s -v --emit-llvm --verify-ir -o=- 2>&1 \
+; RUN:   | FileCheck %s
+
+; ModuleID = 'lgcPipeline'
+source_filename = "lgcPipeline"
+target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7"
+target triple = "amdgcn--amdpal"
+
+; CHECK-LABEL: {{^//}} LLPC location input/output mapping results (FS shader)
+;
+; CHECK:       (FS) Input:  loc = 0, comp = 3 =>  Mapped = 0, 0
+; CHECK-NEXT:  (FS) Input:  loc = 3, comp = 0 =>  Mapped = 0, 1
+; CHECK-NEXT:  (FS) Input:  loc = 3, comp = 1 =>  Mapped = 0, 2
+; CHECK-NEXT:  (FS) Input:  loc = 3, comp = 2 =>  Mapped = 0, 3
+;
+; CHECK:       (FS) Output: loc = 3, comp = 0  =>  Mapped = 0, 0
+
+; CHECK-LABEL: {{^//}} LLPC location input/output mapping results (VS shader)
+;
+; CHECK:       (VS) Input:  loc = 0, comp = 0 =>  Mapped = 0, 0
+; CHECK-NEXT:  (VS) Input:  loc = 1, comp = 0 =>  Mapped = 1, 0
+;
+; CHECK:       (VS) Output: loc = 0, comp = 3  =>  Mapped = 0, 0
+; CHECK-NEXT:  (VS) Output: loc = 3, comp = 0  =>  Mapped = 0, 1
+; CHECK-NEXT:  (VS) Output: loc = 3, comp = 1  =>  Mapped = 0, 2
+; CHECK-NEXT:  (VS) Output: loc = 3, comp = 2  =>  Mapped = 0, 3
+
+; CHECK-LABEL: {{^//}} LLPC pipeline patching results
+
+; Function Attrs: nounwind
+define dllexport spir_func void @lgc.shader.VS.main() local_unnamed_addr #0 !spirv.ExecutionModel !34 !lgc.shaderstage !34 {
+.entry:
+; CHECK:      define dllexport amdgpu_vs void @_amdgpu_vs_main
+; CHECK-NEXT: .entry:
+
+  %a4 = call <3 x float> (...) @lgc.create.read.generic.input.v3f32(i32 1, i32 0, i32 0, i32 0, i32 0, i32 undef)
+  %a5 = call <3 x float> (...) @lgc.create.read.generic.input.v3f32(i32 0, i32 0, i32 0, i32 0, i32 0, i32 undef)
+
+; CHECK: %[[input_load:.+]] = call <3 x i32> @llvm.amdgcn.struct.tbuffer.load.v3i32
+; CHECK: %[[bitcast:.+]] = bitcast <3 x i32> %[[input_load]] to <3 x float>
+
+  %a6 = shufflevector <3 x float> %a5, <3 x float> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 undef>
+  %a7 = insertelement <4 x float> %a6, float 1.000000e+00, i32 3
+
+; CHECK: %[[e0:.+]] = extractelement <3 x float> %[[bitcast]], i32 0
+; CHECK: %[[e1:.+]] = extractelement <3 x float> %[[bitcast]], i32 1
+; CHECK: %[[e2:.+]] = extractelement <3 x float> %[[bitcast]], i32 2
+
+  call void (...) @lgc.create.write.generic.output(<4 x float> %a7, i32 3, i32 0, i32 0, i32 0, i32 0, i32 undef)
+  call void (...) @lgc.create.write.builtin.output(<4 x float> %a7, i32 0, i32 0, i32 undef, i32 undef)
+
+; CHECK: call void @llvm.amdgcn.exp.f32(i32 immarg {{.+}}, i32 immarg {{.+}}, float %[[e0]], float %[[e1]], float %[[e2]], float 1.000000e+00, i1 immarg true, i1 immarg false)
+; CHECK: call void @llvm.amdgcn.exp.f32(i32 {{.+}}, i32 {{.+}}, float undef, float %[[e0]], float %[[e1]], float %[[e2]], i1 false, i1 false)
```

Comment:
This kind of pattern will not be generated in packing mode. The undef value in exp instruction can only be placed at the back of defined value.

Reply:
These are the instructions from the original vertex shader:
```
  %71 = insertelement <4 x float> %66, float %70, i32 3
  %72 = insertelement <4 x float> undef, float %32, i32 0
  %73 = insertelement <4 x float> %72, float %37, i32 1
  %74 = insertelement <4 x float> %73, float %42, i32 2
  call void (...) @lgc.create.write.generic.output(<4 x float> %74, i32 3, i32 0, i32 0, i32 0, i32 0, i32 undef)
  call void (...) @lgc.create.write.builtin.output(<4 x float> %71, i32 0, i32 0, i32 undef, i32 undef)
  call void (...) @lgc.create.write.builtin.output(float undef, i32 1, i32 0, i32 undef, i32 undef)
  call void (...) @lgc.create.write.builtin.output([1 x float] undef, i32 3, i32 4096, i32 undef, i32 undef)
```

Reply:
and from the fragment shader:
```
  %0 = call i8 addrspace(7)* (...) @lgc.create.load.buffer.desc.p7i8(i32 6, i32 4, i32 0, i32 0)
  %1 = call {}* @llvm.invariant.start.p7i8(i64 -1, i8 addrspace(7)* %0)
  %2 = call <4 x float> (...) @lgc.create.read.generic.input.v4f32(i32 0, i32 0, i32 0, i32 0, i32 16, i32 undef)
  %__llpc_input_proxy_.12.vec.extract = extractelement <4 x float> %2, i32 3

...

  %6 = call <4 x float> (...) @lgc.create.read.generic.input.v4f32(i32 3, i32 0, i32 0, i32 0, i32 16, i32 undef)
  %7 = shufflevector <4 x float> %6, <4 x float> undef, <3 x i32> <i32 0, i32 1, i32 2>
  ... ; %9 is based on %7
  %10 = extractelement <3 x float> %9, i32 0
  %11 = extractelement <3 x float> %9, i32 1
  %12 = extractelement <3 x float> %9, i32 2
  %13 = insertelement <4 x float> undef, float %10, i32 0
  %14 = insertelement <4 x float> %13, float %11, i32 1
  %15 = insertelement <4 x float> %14, float %12, i32 2
  ...
    %__llpc_output_proxy_.12.vec.insert = insertelement <4 x float> %15, float %18, i32 3
 ...
     call void (...) @lgc.create.write.generic.output(<4 x float> %__llpc_output_proxy_.0, i32 3, i32 0, i32 0, i32 0, i32 0, i32 undef)
```

Reply:
The mapping in the original pipeline is identical to the mapping from this test.

Reply:
@kuhar Thank you for investigating and fixing it. I got your point. 
Yes, it is a case missing from the original implementation. I had thought a non-vertex input must match an output from a previous shader stage, but it is not true. 
`%2 = call <4 x float> (...) @lgc.create.read.generic.input.v4f32(i32 0, i32 0, i32 0, i32 0, i32 16, i32 undef)` has no matching output in the vertex shader. Although its value is no sense, it is allowed.